### PR TITLE
[TF:TRT] Refactor einsum converter to use IEinsumLayer

### PR DESCRIPTION
**Pull request by tfeher**
This PR refactors the Einsum converter to use IEinsumLayer which is available in TRT 8.2.

Using IEinsumLayer improves the performance and enables conversion of additional equation types.

Additionally the test of validation mode is improved:
 - The expected status of validation was not checked if the actual status was OK.
 - This is now corrected: the validation status is always checked.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc [hunk 1]
```diff
@@ -729,6 +732,192 @@ class ConvertEinsum : public OpConverterBase<ConvertEinsum> {
   std::unique_ptr<EinsumDescriptor> descriptor_a{nullptr};
   std::unique_ptr<EinsumDescriptor> descriptor_b{nullptr};
 };
+#else
+
+// Helper class to reindex equations to contain only lowercase characters.
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:737]: Would you please document these:
   Assumption on the input string, ReIndexer(eq), if there are such assumption.
   How do we reindex equations to contain only lowercase chars.

File: tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc [hunk 2]
```diff
@@ -729,6 +732,192 @@ class ConvertEinsum : public OpConverterBase<ConvertEinsum> {
...
+  // Finds new character for uppercase character c.
+  char operator()(char c) {
+    if (!std::isupper(c)) return c;
+    if (idx_map_.count(c) > 0) return idx_map_[c];
+    char new_idx = std::tolower(c);
+    if (idx_map_.count(new_idx) == 0) {
+      idx_map_[c] = new_idx;
+      idx_map_[new_idx] = new_idx;  // mark that new_idx is taken
+      return new_idx;
+    }
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:757]: Let's add a comment this code block
  // If lower(c) is not used in the equation, use it to replace c.

File: tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc [hunk 3]
```diff
@@ -729,6 +732,192 @@ class ConvertEinsum : public OpConverterBase<ConvertEinsum> {
...
+    for (char k = 'a'; k <= 'z'; k++) {
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:758]: Let's add a comment for this code block:
// Otherwise, find the first available lower case to replace c.

File: tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc [hunk 4]
```diff
@@ -729,6 +732,192 @@ class ConvertEinsum : public OpConverterBase<ConvertEinsum> {
...
+    for (const auto& input : params_->inputs) {
+      VLOG(2) << "input " << (input.is_tensor() ? "tensor" : "weight")
+              << ", dims " << DebugString(input.GetTrtDims());
+    }
+    StatusOr<std::string> eq = GetAttrValue<std::string>("equation");
+    TRT_ENSURE_OK(eq);
+
+    TF_RETURN_IF_ERROR(ValidateEinsumEquation(*eq));
+
+    equation = MakeLowerCase(*eq);
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/ops/einsum.cc:883]: Can you put a comment to explain why we need to rewrite the equation with lower cases?

**Comment by tfeher** [comment]
Thanks @bixia1  for the first look! The PR is still draft, I am debugging a corner case in the unit test, I will re-enable all the tests once that is done. I will address the issues and I will ping you when it is ready for review!

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review, I have addressed the issues so far. The PR is ready for review.

**Review by bixia1 — APPROVED** [review]

**Review by bixia1 — CHANGES_REQUESTED** [review]
Emailed test failure log to author.

**Comment by tfeher** [comment]
As discussed offline with @bixia1, the reason for the failure was a modification of the validation checks. I have separated that change into #56843. 

The tests for this PR should be passing now. We can merge the current PR first, and continue with #56843 afterwards.

**Review by bixia1 — APPROVED** [review]
